### PR TITLE
Add expanding channels for conv2d_nhwc and post ops vectorization

### PR DESCRIPTION
**Pull request by echuraev**

### CODE REVIEW

**Comment by echuraev** [comment]
@TejashShah please review this PR

File: tvm/python/tvm/topi/adreno/conv2d_nhwc.py [hunk 1]
```diff
@@ -186,38 +242,34 @@ def schedule_conv2d_NHWC(cfg, s, output, args={}):
...
-    s[kernel].compute_inline()
+    pack_data = pad_data.op.input_tensors[0]
+    if s[pack_data].op.tag != "input_pack_expanded":
+        s[pack_data].compute_inline()
+    else:
+        axes = s[pack_data].op.axis
+        fused = s[pack_data].fuse(*axes[:-1])
+        shape = get_const_tuple(pack_data.shape)
+        ftc = numpy.prod(shape[:-1])
+        div = getDiv(ftc, 64)
```
Comment (TejashShah) [inline tvm/python/tvm/topi/adreno/conv2d_nhwc.py:254]: could you please provide comment on the reason for choosing 64? Is it workgroup or warp size?

File: tvm/python/tvm/topi/adreno/conv2d_nhwc.py [hunk 2]
```diff
@@ -281,17 +333,31 @@ def copy_to_texture(stage):
...
-        s[latest].bind(thread, te.thread_axis("threadIdx.x"))
+    if OC % 4 == 0:
+        n, oh, ow, oc = s[latest].op.axis
+        occ, ocb = s[latest].split(oc, factor=4)
+        s[latest].reorder(n, oh, ow, occ, ocb)
+        s[latest].vectorize(ocb)
+        fused = s[latest].fuse(n, oh, ow, occ)
+
+        ftc = N * OH * OW * OC / 4
+        div = getDiv(ftc, 128)
```
Comment (TejashShah) [inline tvm/python/tvm/topi/adreno/conv2d_nhwc.py:346]: could you please provide comment on the reason for choosing 128? Is it workgroup or warp size?

File: tvm/python/tvm/topi/adreno/conv2d_nhwc.py [hunk 3]
```diff
@@ -281,17 +333,31 @@ def copy_to_texture(stage):
...
 
-    if OC < 32:
-        block, thread = s[latest].split(fused, factor=32)
-        s[latest].bind(block, te.thread_axis("blockIdx.x"))
-        s[latest].bind(thread, te.thread_axis("threadIdx.x"))
+    if OC % 4 == 0:
+        n, oh, ow, oc = s[latest].op.axis
+        occ, ocb = s[latest].split(oc, factor=4)
+        s[latest].reorder(n, oh, ow, occ, ocb)
+        s[latest].vectorize(ocb)
```
Comment (TejashShah) [inline tvm/python/tvm/topi/adreno/conv2d_nhwc.py:342]: Just out of curiosity, does this vectorization apply to just rewrites of conv operation or of any operation fused into conv operation as well?

For e.g. if relu is fused into conv (conv+relu), does this vectorization apply to relu rewrites?

**Review by TejashShah — CHANGES_REQUESTED** [review]
Please add comments. Otherwise LGTM.

File: tvm/python/tvm/topi/adreno/conv2d_nhwc.py (hunk 1 above)
Comment (elvin-n) [reply tvm/python/tvm/topi/adreno/conv2d_nhwc.py:254]: By fact this parameter must be a subject for tuning as well. Unfortunately adding three more parameters to the current schedules will increase search space even more. Even right now there are millions combinations to measure. If we add 3 more parameters each of them can be represented by even several numbers, the seach space might be increased to several hundreds millions variants.
So far I decided not to add it to the search space but hard code them based on my manual measurements. On several standalone convolutions I saw that 64 is the best choice in this certain place (for repacking of inputs). In other place I saw that 128 is the best choice.

File: tvm/python/tvm/topi/adreno/conv2d_nhwc.py (hunk 2 above)
Comment (elvin-n) [reply tvm/python/tvm/topi/adreno/conv2d_nhwc.py:346]: see above comment

File: tvm/python/tvm/topi/adreno/conv2d_nhwc.py (hunk 3 above)
Comment (elvin-n) [reply tvm/python/tvm/topi/adreno/conv2d_nhwc.py:342]: it is applied for all post-ops (been fused to one op and scheduled in this place)

**Comment by TejashShah** [comment]
LGTM

cc @tmoreau89

**Pull request merged**